From 945a40edf0f3590b265e50177cc65527a2994c58 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Wed, 30 May 2018 10:19:20 +0800
Subject: [PATCH 0654/1051] arm64: dts: marvell: 8040: add configs for setups B
 C D E

8040 DB board can have different setups for different kinds of
usage by changing the DPR and switches on board.
The detailed differences of these setups can be found in:
Documentation/mvebu/a7k-a8k/armada8040-db-setup.txt
The TDM and slic driver are still missing in the current kernel
source tree, should add dts support after adding these drivers

Change-Id: Ib47e3559ff96f9d308ef5193c2582d95115b3b50
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../bindings/arm/marvell/armada-7k-8k.txt     |  12 +
 .../mvebu/a7k-a8k/armada8040-db-setup.txt     | 175 +++++++++++++++
 arch/arm64/boot/dts/marvell/Makefile          |   4 +
 .../boot/dts/marvell/armada-8040-db-B.dts     | 153 +++++++++++++
 .../boot/dts/marvell/armada-8040-db-C.dts     | 170 +++++++++++++++
 .../boot/dts/marvell/armada-8040-db-D.dts     | 177 +++++++++++++++
 .../boot/dts/marvell/armada-8040-db-E.dts     | 205 ++++++++++++++++++
 .../boot/dts/marvell/armada-8040-db.dtsi      | 195 +++++++++++++++++
 8 files changed, 1091 insertions(+)
 create mode 100644 Documentation/mvebu/a7k-a8k/armada8040-db-setup.txt
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db-B.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db-C.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db-D.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db-E.dts
 create mode 100644 arch/arm64/boot/dts/marvell/armada-8040-db.dtsi

diff --git a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
index a516c4ff3322..affaa3376ead 100644
--- a/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
+++ b/Documentation/devicetree/bindings/arm/marvell/armada-7k-8k.txt
@@ -37,6 +37,18 @@ Boards:
    - "marvell,armada8040-db-default", "marvell,armada8040-db"
       when the board being used is the Armada 8040 Development board
 
+   - "marvell,armada8040-db-B", "marvell,armada8040-db"
+      when the board being used is the Armada 8040 Development board with Setup B
+
+   - "marvell,armada8040-db-C", "marvell,armada8040-db"
+      when the board being used is the Armada 8040 Development board with Setup C
+
+   - "marvell,armada8040-db-D", "marvell,armada8040-db"
+      when the board being used is the Armada 8040 Development board with Setup D
+
+   - "marvell,armada8040-db-E", "marvell,armada8040-db"
+      when the board being used is the Armada 8040 Development board with Setup E
+
    - "marvell,armada8040-mcbin"
       when the board being used is the MACCHIATOBin community board
 
diff --git a/Documentation/mvebu/a7k-a8k/armada8040-db-setup.txt b/Documentation/mvebu/a7k-a8k/armada8040-db-setup.txt
new file mode 100644
index 000000000000..0f03905aa1e2
--- /dev/null
+++ b/Documentation/mvebu/a7k-a8k/armada8040-db-setup.txt
@@ -0,0 +1,175 @@
+Armada8040 development board setup:
+--------------------------------------
+
+The Armada 8040 development board (PCB: DB-88F8040-Modular) is a configurable board,
+designed to test several combinations of SoC configurations.
+linux supports a defined set of those configurations via different device tree files.
+
+This document describes the board modifications required to set up each configuration and the
+interfaces supported (or disabled in each one).
+
+To switch board configuration:
+1. Modify HW board settings according to the instructions in the "Board Setup" section below.
+2. Use appropriate Device Tree file for booting Linux.
+3. Same Device Tree selection is required also in U-Boot (to initialize requested comphy SerDes lanes).
+   Modify the Device Tree file used by U-BOOT, by running the U-BOOT command "fdt_config",
+   and select the appropriate configuration entry.
+
+
+Board Setup
+------------
+
+For A8040 A1 revision SoC - Please use armada-8040-db configuration,
+			    Please use armada-8040-db-B configuration,
+			    Please use armada-8040-db-C configuration,
+			    Please use armada-8040-db-D configuration,
+			    Please use armada-8040-db-E configuration
+
+|		| 1: A			| 2: B			| 3: C			| 4: D			| 5: E			|
+|---------------|-----------------------|-----------------------|-----------------------|-----------------------|-----------------------|
+|Device tree	|armada-8040-db.dtb	|armada-8040-db-B.dtb	|armada-8040-db-C.dtb	|armada-8040-db-D.dtb	|armada-8040-db-E.dtb	|
+|CP0-SW1:1	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP0-SW1:2	| OFF			| ON			| OFF			| OFF			| OFF			|
+|CP0-SW1:3	| OFF			| ON			| OFF			| OFF			| OFF			|
+|CP0-SW1:4	| OFF			| ON			| OFF			| OFF			| OFF			|
+|CP0-SW1:5	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP0-SW1:6	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP0-SW1:7	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP0-SW1:8	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:1	| OFF			| ON			| OFF			| OFF			| OFF			|
+|CP1-SW1:2	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:3	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:4	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:5	| ON			| ON			| ON			| ON			| ON			|
+|CP1-SW1:6	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:7	| OFF			| OFF			| OFF			| OFF			| OFF			|
+|CP1-SW1:8	| OFF			| OFF			| OFF			| OFF			| OFF			|
+
+DPR Setting
+------------
+|						| 1: A		| 2: B			| 3: C						|
+|-----------------------------------------------|---------------|-----------------------|-----------------------------------------------|
+| Jumper-module (CON44): SLM1564		| DPR all = 1-2	| DPR all = 1-2		| DPR 96-101, 103-108 = 2-3, DPR 102 = 1-2	|
+| Serdes CP0 site module (CON42): SLM1521	| DPR 2-5 = 1-2	| DPR 2-5 = 2-3		| DPR 2-5 = 1-2					|
+| Serdes CP1 site module (CON41): SLM1521	| DPR 2-5 = 1-2	| DPR 2-5 = 1-2		| DPR 2-5 = 1-2					|
+| DB8040					|		|			| DPR 84 = 2-3					|
+|						|		|			| TDM module (CON51): SLM1448			|
+
+|						| 4: D						| 5: E					|
+|-----------------------------------------------|-----------------------------------------------|---------------------------------------|
+| Jumper-module (CON44): SLM1564		| DPR 96-101,103-108 = 2-3, DPR 102 = 1-2	| DPR 109-112 = 2-3, others = 1-2	|
+| Serdes CP0 site module (CON42): SLM1521	| DPR 2-5 = 1-2					| DPR 2-5 = 1-2				|
+| Serdes CP1 site module (CON41): SLM1521	| DPR 2-5 = 1-2					| DPR 2-5 = 1-2				|
+| DB8040					| DPR 76 = 1-2, DPR 87-88 = 2-3			| DPR 81 = 2-3, DPR 23-26,119,121 =1-2	|
+|						| Audio module (CON51): SLM1449			|					|
+
+The tables below summarizes the interface configuration of each setup
+
+SerDes Configuration
+------------------------
+
+ CP0 Lane	| 1: A		| 2: B		| 3: C		| 4: D		| 5: E		|
+----------------|---------------|---------------|---------------|---------------|---------------|
+ 0		| PCIe0 (x1)	| PCIe0 (x4)	| PCIe0 (x1)	| PCIe0 (x1)	| PCIe0 (x1)	|
+ 1		| SATA0		| PCIe0 (x4)	| SATA0		| SATA0		| SATA0		|
+ 2		| SFI (10G)	| PCIe0 (x4)	| SFI (10G)	| SFI (10G)	| SFI (10G)	|
+ 3		| SATA1		| PCIe0 (x4)	| SATA1		| SATA1		| SATA1		|
+ 4		| USB3_HOST1	| USB3_HOST1	| USB3_HOST1	| USB3_HOST1	| USB3_HOST1	|
+ 5		| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	|
+
+- USB2_H0 (UTMI only)
+- USB2_H1 (UTMI & USB3 SerDes)
+
+ CP1 Lane	| 1: A		| 2: B		| 3: C		| 4: D		| 5: E		|
+----------------|---------------|---------------|---------------|---------------|---------------|
+ 0		| PCIe0 (x1)	| PCIe0 (x1)	| PCIe0 (x1)	| PCIe0 (x1)	| PCIe0 (x1)	|
+ 1		| SATA0		| USB3_HOST0	| SATA0		| SATA0		| SATA0		|
+ 2		| SFI (10G)	| SFI (10G)	| SFI (10G)	| SGMII0	| SGMII0 (2.5G)	|
+ 3		| SATA1		| SATA1		| SATA1		| SATA1		| SATA1		|
+ 4		| PCIe1 (x1)	| PCIe1 (x1)	| PCIe1 (x1)	| PCIe1 (x1)	| PCIe1 (x1)	|
+ 5		| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	| PCIe2 (x1)	|
+
+- USB2_H0 (UTMI only)
+
+
+Multi-purpose pin configurations
+--------------------------------
+
+ AP806 pin	| 1: A		| 2: B		| 3: C		| 4: D		| 5: E		|
+----------------|---------------|---------------|---------------|---------------|---------------|
+    AP-SDIO	| [0-10]	| [0-10]	| [0-10]	| [0-10]	| N/C		|
+    I2C0	| N/C		| N/C		| N/C		| N/C		| N/C		|
+    AP-SPI0	| N/C		| N/C		| N/C		| N/C		| [0,3]		|
+    AP-UART0	| [11,19]	| [11,19]	| [11,19]	| [11,19]	| [11,19]	|
+
+ CP0 pin	| 1: A		| 2: B		| 3: C		| 4: D		| 5: E		|
+----------------|---------------|---------------|---------------|---------------|---------------|
+    CP0-SMI	| [32,34]	| [32,34]	| [32,34]	| [32,34]	| N/C		|
+    CP0-XSMI	| [42-43]	| [42-43]	| [42-43]	| [42-43]	| [42-43]	|
+    CP0-TWSI	| [37-38]	| [37-38]	| [37-38]	| [37-38]	| [37-38]	|
+    CP0-SATA1	| [40]		| [40]		| [40]		| [40]		| [40]		|
+    CP0-SATA0	| [41]		| [41]		| [41]		| [41]		| [41]		|
+    CP0-RGMII1	| [44-55]	| [44-55]	| [44-55]	| [44-55]	| N/C		|
+    CP0-SD	| [56-62]	| [56-62]	| [56-62]	| [56-62]	| [56-62]	|
+    CP0-PTP_CLK	| [39]		| [39]		| [39]		| [39]		| [39]		|
+    CP0-TWSI1	| N/C		| N/C		| N/C		| N/C		| [35-36]	|
+    CP0-UART1	| N/C		| N/C		| N/C		| N/C		| [47,49]	|
+    CP0-UART2	| N/C		| N/C		| N/C		| N/C		| [50-51]	|
+
+GPIOS:
+	- 33: GPIO: GE_INT#/push button/Wake
+	- 35: MSS_GPIO[3]: MSS_PWDN (only in configuration 1,3,4,5)
+	- 36: MSS_GPIO[5]: MSS_VTT_EN (only in configuration 1,3,4,5)
+
+ CP1 pin	| 1: A		| 2: B		| 3: C		| 4: D		| 5: E		|
+----------------|---------------|---------------|---------------|---------------|---------------|
+    CP1-RGMII0	| [0-11]	| [0-11]	| N/C		| N/C		| [0-11]	|
+    CP1-SMI	| [27,31]	| [27,31]	| N/C		| N/C		| [27,31]	|
+    CP1-SPI1	| [13-16]	| [13-16]	| [13-16]	| N/C		| [13-16]	|
+    CP1-UART0	| [29-30]	| [29-30]	| [29-30]	| [29-30]	| [29-30]	|
+    CP1-SATA1	| [28]		| [28]		| [28]		| [28]		| [28]		|
+    CP1-SATA0	| N/C		| N/C		| [27]		| N/C		| N/C		|
+    CP1-AU	| N/C		| N/C		| N/C		| [0-5]		| N/C		|
+    CP1-NAND	| N/C		| N/C		| N/C		| [13,15-27]	| N/C		|
+    CP1-TDM	| N/C		| N/C		| [0-11]	| N/C		| N/C		|
+
+Network configuration
+---------------------
+1: A:
+ Interface	| CPN#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| 0	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+ eth1		| 0	| 2		| 3		| RGMII1			|
+ eth2		| 1	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+ eth3		| 1	| 1		| 2		| RGMII0			|
+
+2: B:
+ Interface	| CPN#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| 0	| 2		| 3		| RGMII1			|
+ eth1		| 1	| 0		| 0		| SFI (10G Lane 2 connected to SFP)
+ eth2		| 1	| 1		| 2		| RGMII0			|
+
+3: C:
+ Interface	| CPN#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| 0	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+ eth1		| 0	| 2		| 3		| RGMII1			|
+ eth2		| 1	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+
+4: D:
+ Interface	| CPN#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| 0	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+ eth1		| 0	| 2		| 3		| RGMII1			|
+ eth2		| 1	| 0		| 0		| SGMII0			|
+
+5: E:
+ Interface	| CPN#	| PPv2 Port	| GMAC Port	| Board Interface		|
+ ---------------|-------|---------------|---------------|-------------------------------|
+ eth0		| 0	| 0		| 0		| 10G (Lane 2 connected to SFP)	|
+ eth1		| 1	| 0		| 0		| 1000base-x HS_SGMII0, speed=2500
+ eth2		| 1	| 1		| 2		| RGMII0			|
+
+Notes:
+	- eth0/1/2/3 may in fact be higher numbers, if prior eth_x interfaces already exist.
diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index 9bb489fd6648..ce027931b0b8 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -8,6 +8,10 @@ dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-C.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-D.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-E.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db-B.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db-C.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db-D.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-db-E.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8040-mcbin-single-shot.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-8080-pd.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db-B.dts b/arch/arm64/boot/dts/marvell/armada-8040-db-B.dts
new file mode 100644
index 000000000000..8328add192d0
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db-B.dts
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 8040 development board
+ * This file supports option B configuration
+ * AP_UART,AP_eMMC,CP0_I2C,CP0_eMMC,CP0_RGMII1,CP1_SPI(Boot),CP1_RGMII0
+ */
+
+#include "armada-8040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board B setup";
+	compatible = "marvell,armada8040-db-B", "marvell,armada8040-db",
+		"marvell,armada8040", "marvell,armada-ap806-quad",
+		"marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth2;
+		ethernet1 = &cp1_eth0;
+		ethernet2 = &cp1_eth1;
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+/* CON6 on CP0 expansion */
+&cp0_pcie0 {
+	num-lanes = <4>;
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP0 expansion */
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON9 on CP0 expansion */
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+/* CON10 on CP0 expansion */
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&phy1>;
+	phy-mode = "rgmii-id";
+};
+
+&cp1_spi1 {
+	status = "okay";
+};
+
+/* CON6 on CP1 expansion */
+&cp1_pcie0 {
+	status = "okay";
+	phys = <&cp1_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON7 on CP1 expansion */
+&cp1_pcie1 {
+	status = "okay";
+	phys = <&cp1_comphy4 1>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP1 expansion */
+&cp1_pcie2 {
+	status = "okay";
+	phys = <&cp1_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON4 on CP1 expansion */
+&cp1_sata0 {
+	status = "okay";
+
+	/* CON4 on CP1 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy3 1>;
+	};
+};
+
+/* CON9 on CP1 expansion */
+&cp1_usb3_0 {
+	status = "okay";
+};
+
+&cp1_mdio {
+	status = "okay";
+
+	cp1_phy1: ethernet-phy@2 {
+		reg = <0>;
+	};
+};
+
+&cp1_ethernet {
+	status = "okay";
+};
+
+&cp1_eth0 {
+	status = "okay";
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp1_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth1>;
+};
+
+&cp1_eth1 {
+	status = "okay";
+	phy = <&cp1_phy1>;
+	phy-mode = "rgmii-id";
+};
+
+&cp0_crypto {
+	status = "okay";
+};
+
+&cp1_crypto {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db-C.dts b/arch/arm64/boot/dts/marvell/armada-8040-db-C.dts
new file mode 100644
index 000000000000..621f84880717
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db-C.dts
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 8040 development board
+ * This file supports option C configuration
+ * AP_UART,AP_eMMC-8b,TDM,CP0_SDIO-4b,CP0_I2C,CP0_eMMC,CP0_RGMII1,
+ * CP1_SPI(Boot),CP1_RGMII0
+ */
+
+#include "armada-8040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board C setup";
+	compatible = "marvell,armada8040-db-C", "marvell,armada8040-db",
+		"marvell,armada8040", "marvell,armada-ap806-quad",
+		"marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp0_eth2;
+		ethernet2 = &cp1_eth0;
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+/* CON6 on CP0 expansion */
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP0 expansion */
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+	/* CON2 on CP0 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy1 0>;
+	};
+	/* CON4 on CP0 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+/* CON9 on CP0 expansion */
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+/* CON10 on CP0 expansion */
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth0>;
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&phy1>;
+	phy-mode = "rgmii-id";
+};
+
+&cp1_spi1 {
+	status = "okay";
+};
+
+/* CON6 on CP1 expansion */
+&cp1_pcie0 {
+	status = "okay";
+	phys = <&cp1_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON7 on CP1 expansion */
+&cp1_pcie1 {
+	status = "okay";
+	phys = <&cp1_comphy4 1>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP1 expansion */
+&cp1_pcie2 {
+	status = "okay";
+	phys = <&cp1_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON4 on CP1 expansion */
+&cp1_sata0 {
+	status = "okay";
+	/* CON2 on CP1 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy1 0>;
+	};
+	/* CON4 on CP1 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy3 1>;
+	};
+};
+
+/* CON9 on CP1 expansion */
+&cp1_usb3_0 {
+	usb-phy = <&cp1_usb3_0_phy>;
+	status = "okay";
+};
+
+&cp1_ethernet {
+	status = "okay";
+};
+
+&cp1_eth0 {
+	status = "okay";
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp1_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth1>;
+};
+
+&cp0_crypto {
+	status = "okay";
+};
+
+&cp1_crypto {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db-D.dts b/arch/arm64/boot/dts/marvell/armada-8040-db-D.dts
new file mode 100644
index 000000000000..b4c75650fa4e
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db-D.dts
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 8040 development board
+ * This file supports option D configuration
+ * AP_UART,AP_eMMC-8b,CP0_RGMII1,CP0_SDIO-4b,CP0_I2C,CP1_NAND,CP1_UART
+ */
+
+#include "armada-8040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board D setup";
+	compatible = "marvell,armada8040-db-D", "marvell,armada8040-db",
+		"marvell,armada8040", "marvell,armada-ap806-quad",
+		"marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp0_eth2;
+		ethernet2 = &cp1_eth0;
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+/* CON6 on CP0 expansion */
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP0 expansion */
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+	/* CON2 on CP0 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy1 0>;
+	};
+	/* CON4 on CP0 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+/* CON9 on CP0 expansion */
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+/* CON10 on CP0 expansion */
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_mdio {
+	status = "okay";
+
+	phy1: ethernet-phy@1 {
+		reg = <1>;
+	};
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth0>;
+};
+
+&cp0_eth2 {
+	status = "okay";
+	phy = <&phy1>;
+	phy-mode = "rgmii-id";
+};
+
+
+&cp1_nand {
+	status = "okay";
+};
+
+/* CON6 on CP1 expansion */
+&cp1_pcie0 {
+	status = "okay";
+	phys = <&cp1_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON7 on CP1 expansion */
+&cp1_pcie1 {
+	status = "okay";
+	phys = <&cp1_comphy4 1>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP1 expansion */
+&cp1_pcie2 {
+	status = "okay";
+	phys = <&cp1_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON4 on CP1 expansion */
+&cp1_sata0 {
+	status = "okay";
+	/* CON2 on CP1 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy1 0>;
+	};
+	/* CON4 on CP1 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy3 1>;
+	};
+};
+
+/* CON9 on CP1 expansion */
+&cp1_usb3_0 {
+	status = "okay";
+};
+
+&cp1_mdio {
+	status = "okay";
+
+	cp1_phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
+
+&cp1_ethernet {
+	status = "okay";
+};
+
+&cp1_eth0 {
+	status = "okay";
+	/* Network PHY */
+	phy = <&cp1_phy0>;
+	phy-mode = "sgmii";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp1_comphy2 0>;
+};
+
+&cp0_crypto {
+	status = "okay";
+};
+
+&cp1_crypto {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db-E.dts b/arch/arm64/boot/dts/marvell/armada-8040-db-E.dts
new file mode 100644
index 000000000000..0d24ef43de3b
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db-E.dts
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 8040 development board
+ * This file supports option default (A) configuration
+ * 4xUART(AP,CP1_UA0,2xCP0_UA0),AP_I2C,AP_SPI,CP0_I2C0,CP0_I2C1,
+ * CP0_SDIO-4b,CP1_SPI,CP1_RGMII0
+ */
+
+#include "armada-8040-db.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board E setup";
+	compatible = "marvell,armada8040-db-E", "marvell,armada8040-db",
+		"marvell,armada8040", "marvell,armada-ap806-quad",
+		"marvell,armada-ap806";
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	aliases {
+		ethernet0 = &cp0_eth0;
+		ethernet1 = &cp1_eth0;
+		ethernet2 = &cp1_eth1;
+	};
+
+	ap_pinctrl {
+		ap_spi0_pins: spi-pins-0 {
+			marvell,pins = "mpp0", "mpp1", "mpp2",
+			"mpp3";
+			marvell,function = "spi0";
+		};
+	};
+};
+
+&ap_sdhci0 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&ap_spi0_pins>;
+};
+
+&cp0_i2c1 {
+	status = "okay";
+	clock-frequency = <100000>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&cp0_i2c1_pins>;
+};
+
+&cp0_sdhci0 {
+	status = "okay";
+};
+
+/* CON6 on CP0 expansion */
+&cp0_pcie0 {
+	status = "okay";
+	phys = <&cp0_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP0 expansion */
+&cp0_pcie2 {
+	status = "okay";
+	phys = <&cp0_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+&cp0_sata0 {
+	status = "okay";
+	/* CON2 on CP0 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy1 0>;
+	};
+	/* CON4 on CP0 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp0_comphy3 1>;
+	};
+};
+
+&cp0_pinctrl {
+	cp0_i2c1_pins: i2c-pins-1 {
+		marvell,pins = "mpp35", "mpp36";
+		marvell,function = "i2c1";
+	};
+};
+
+&cp0_serial0 {
+	status = "okay";
+};
+
+&cp0_serial1 {
+	status = "okay";
+};
+
+/* CON9 on CP0 expansion */
+&cp0_usb3_0 {
+	status = "okay";
+};
+
+/* CON10 on CP0 expansion */
+&cp0_usb3_1 {
+	status = "okay";
+};
+
+&cp0_ethernet {
+	status = "okay";
+};
+
+&cp0_eth0 {
+	status = "okay";
+	phy-mode = "10gbase-kr";
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy2 0>;
+	managed = "in-band-status";
+	sfp = <&sfp_eth0>;
+};
+
+&cp1_spi1 {
+	status = "okay";
+};
+
+/* CON6 on CP1 expansion */
+&cp1_pcie0 {
+	status = "okay";
+	phys = <&cp1_comphy0 0>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON7 on CP1 expansion */
+&cp1_pcie1 {
+	status = "okay";
+	phys = <&cp1_comphy4 1>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON5 on CP1 expansion */
+&cp1_pcie2 {
+	status = "okay";
+	phys = <&cp1_comphy5 2>;
+	phy-names = "pcie-phy0";
+};
+
+/* CON4 on CP1 expansion */
+&cp1_sata0 {
+	status = "okay";
+	/* CON2 on CP1 expansion */
+	sata-port@0 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy1 0>;
+	};
+	/* CON4 on CP1 expansion */
+	sata-port@1 {
+		status = "okay";
+		/* Generic PHY, providing serdes lanes */
+		phys = <&cp1_comphy3 1>;
+	};
+};
+
+/* CON9 on CP1 expansion */
+&cp1_usb3_0 {
+	status = "okay";
+};
+
+&cp1_mdio {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
+
+&cp1_ethernet {
+	status = "okay";
+};
+
+&cp1_eth0 {
+	status = "okay";
+	/* Network PHY */
+	phy-mode = "sgmii";
+	phy-speed = <2500>;
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp1_comphy2 0>;
+};
+
+&cp1_eth1 {
+	status = "okay";
+	phy = <&phy0>;
+	phy-mode = "rgmii-id";
+};
+
+&cp0_crypto {
+	status = "okay";
+};
+
+&cp1_crypto {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi b/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi
new file mode 100644
index 000000000000..7d5e82d30674
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-8040-db.dtsi
@@ -0,0 +1,195 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Device Tree file for Marvell Armada 8040 development board platform
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+#include "armada-8040.dtsi"
+
+/ {
+	model = "Marvell Armada 8040 development board";
+	compatible = "marvell,armada8040-db", "marvell,armada8040",
+		"marvell,armada-ap806-quad", "marvell,armada-ap806";
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x0 0x0 0x80000000>;
+	};
+
+	cp0_reg_usb3_0_vbus: cp0-usb3-0-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "cp0-usb3h0-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		gpio = <&expander0 0 GPIO_ACTIVE_HIGH>;
+	};
+
+	cp0_reg_usb3_1_vbus: cp0-usb3-1-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "cp0-usb3h1-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		gpio = <&expander0 1 GPIO_ACTIVE_HIGH>;
+	};
+
+	cp0_usb3_0_phy: cp0-usb3-0-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_0_vbus>;
+	};
+
+	cp0_usb3_1_phy: cp0-usb3-1-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp0_reg_usb3_1_vbus>;
+	};
+
+	cp1_reg_usb3_0_vbus: cp1-usb3-0-vbus {
+		compatible = "regulator-fixed";
+		regulator-name = "cp1-usb3h0-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		enable-active-high;
+		gpio = <&expander1 0 GPIO_ACTIVE_HIGH>;
+	};
+
+	cp1_usb3_0_phy: cp1-usb3-0-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&cp1_reg_usb3_0_vbus>;
+	};
+
+	sfp_eth0: sfp-eth0 {
+		compatible = "sff,sfp";
+	};
+
+	sfp_eth1: sfp-eth1 {
+		compatible = "sff,sfp";
+	};
+};
+
+/* Accessible over the mini-USB CON9 connector on the main board */
+&uart0 {
+	status = "okay";
+	pinctrl-0 = <&uart0_pins>;
+	pinctrl-names = "default";
+};
+
+&ap_sdhci0 {
+	bus-width = <8>;
+	non-removable;
+};
+
+&spi0 {
+	status = "disabled";
+	spi-flash@0 {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+		compatible = "jedec,spi-nor";
+		reg = <0x0>;
+		spi-max-frequency = <20000000>;
+		partition@0 {
+			label = "boot";
+			reg = <0x0 0x200000>;
+		};
+		partition@200000 {
+			label = "recovery_image";
+			reg = <0x200000 0x200000>;
+		};
+		partition@400000 {
+			label = "Filesystem";
+			reg = <0x400000 0xc00000>;
+		};
+	};
+};
+
+&cp0_i2c0 {
+	status = "okay";
+	clock-frequency = <100000>;
+
+	/* U31 */
+	expander0: pca9555@21 {
+		compatible = "nxp,pca9555";
+		pinctrl-names = "default";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x21>;
+	};
+
+	/* U25 */
+	expander1: pca9555@25 {
+		compatible = "nxp,pca9555";
+		pinctrl-names = "default";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x25>;
+	};
+
+};
+
+&cp0_sdhci0 {
+	bus-width = <4>;
+	non-removable;
+};
+
+/* CON9 on CP0 expansion */
+&cp0_usb3_0 {
+	usb-phy = <&cp0_usb3_0_phy>;
+};
+
+/* CON10 on CP0 expansion */
+&cp0_usb3_1 {
+	usb-phy = <&cp0_usb3_1_phy>;
+	/* Generic PHY, providing serdes lanes */
+	phys = <&cp0_comphy4 1>;
+	phy-names = "usb";
+};
+
+&cp1_spi1 {
+	status = "disabled";
+
+	spi-flash@0 {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+		compatible = "jedec,spi-nor";
+		reg = <0x0>;
+		spi-max-frequency = <20000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "Boot";
+				reg = <0x0 0x400000>;
+			};
+			partition@400000 {
+				label = "Filesystem";
+				reg = <0x400000 0xc00000>;
+			};
+		};
+	};
+};
+
+&cp1_nand_controller {
+	nand-ecc-strength = <4>;
+	nand-ecc-step-size = <512>;
+	status = "disabled";
+	partition@0 {
+		label = "U-Boot";
+		reg = <0 0x200000>;
+	};
+	partition@200000 {
+		label = "Linux";
+		reg = <0x200000 0xd00000>;
+	};
+	partition@1000000 {
+		label = "Filesystem";
+		reg = <0x1000000 0x3f000000>;
+	};
+};
+
+/* CON9 on CP1 expansion */
+&cp1_usb3_0 {
+	usb-phy = <&cp1_usb3_0_phy>;
+};
-- 
2.17.1

